pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+        #[pallet::weight(T::WeightInfo::nominator_nominate())]
+        pub fn nominator_nominate(
+            origin: OriginFor<T>,
+            validator: T::AccountId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let nominator_acc = ensure_signed(origin)?;
+
+            // cannot be a validator candidate and nominator with same AccountId
+            ensure!(
+                !Self::is_validator(&nominator_acc),
+                Error::<T>::ValidatorExists
+            );
+
+            ensure!(
+                amount >= T::MinNomination::get(),
+                Error::<T>::NominationBelowMin
+            );
+
+            let mut validator_state =
+                <ValidatorState<T>>::get(&validator).ok_or(Error::<T>::ValidatorDNE)?;
+
+            let mut do_add_nomination = true;
+            let mut nominator_state = if Self::is_nominator(&nominator_acc) {
+                Self::nominator_state(&nominator_acc).ok_or(Error::<T>::NominatorDNE)?
+            } else {
+                do_add_nomination = false;
+                Nominator::new(validator.clone(), amount)
+            };
+
+            ensure!(
+                (nominator_state.nominations.0.len() as u32) < T::MaxValidatorPerNominator::get(),
+                Error::<T>::ExceedMaxValidatorPerNom,
+            );
+
+            ensure!(
+                (validator_state.nominators.0.len() as u32) < T::MaxNominatorsPerValidator::get(),
+                Error::<T>::TooManyNominators,
+            );
+
+            let nominator_free_balance = T::Currency::free_balance(&nominator_acc);
+
+            if do_add_nomination {
+                ensure!(
+                    nominator_free_balance >= nominator_state.total + amount,
+                    Error::<T>::InsufficientBalance
+                );
+                ensure!(
+                    nominator_state.add_nomination(Bond {
+                        owner: validator.clone(),
+                        amount
+                    }),
+                    Error::<T>::AlreadyNominatedValidator,
+                );
+            } else {
+                ensure!(
+                    nominator_free_balance >= amount,
+                    Error::<T>::InsufficientBalance
+                );
+            }
+
+            let nomination = Bond {
+                owner: nominator_acc.clone(),
+                amount,
+            };
+            ensure!(
+                validator_state.nominators.insert(nomination),
+                Error::<T>::NominatorExists,
+            );
+
+            // T::Currency::reserve(&nominator_acc, amount)?;
+            T::Currency::set_lock(
+                STAKING_ID,
+                &nominator_acc,
+                nominator_state.total,
+                WithdrawReasons::all(),
+            );
+
+            validator_state.nomi_bond_total += amount;
+            validator_state.total += amount;
```

Comment:
use safe math or saturating operators :)

Reply:
:+1: , Moved to saturating_add()